### PR TITLE
Feature / Lock - Unlock - Draft revisions

### PR DESCRIPTION
**Pull request by coppee**
Needed to the "Hippoc client demo" to avoid some errors in the publish quarter

### CODE REVIEW

File: Command/TransformContentTypeCommand.php [hunk 1]
```diff
@@ -123,4 +133,26 @@ private function setContentTypeNameArgument(InputInterface $input, string $messa
...
         $input->setArgument(self::ARGUMENT_CONTENTTYPE_NAME, $contentTypeName);
     }
+
+    private function checkUserArgument(InputInterface $input): void
+    {
+        $user = $input->getArgument(self::ARGUMENT_USER);
+        if (null === $user) {
+            $message = 'The user name is not provided';
+            $this->setUserArgument($input, $message);
+            return;
```
Comment (Davidmattei) [inline Command/TransformContentTypeCommand.php]:
```suggestion
```
The return can be removed

File: Command/TransformContentTypeCommand.php [hunk 2]
```diff
@@ -95,24 +104,25 @@ protected function execute(InputInterface $input, OutputInterface $output)
         return 0;
     }
 
-    private function checkContentTypeNameArgument(InputInterface $input)
+    private function checkContentTypeArgument(InputInterface $input)
```
Comment (Davidmattei) [inline Command/TransformContentTypeCommand.php]: add return type :void

File: Repository/RevisionRepository.php [hunk 3]
```diff
@@ -609,6 +609,34 @@ public function lockRevisions(ContentType $contentType, \DateTime $until, $by, $
         return $qbUpdate->getQuery()->execute();
     }
 
+    public function unlockContentTypeRevisionsbyUser(ContentType $contentType, string $by): int
+    {
+        $params = ['content_type' => $contentType, 'by' => $by, 'null' => null];
+
+        /** @var QueryBuilder $qbSelect */
```
Comment (Davidmattei) [inline Repository/RevisionRepository.php]: please remove, $this->createQueryBuilder() has a correct doc block, so this is overkill :)

File: Repository/RevisionRepository.php [hunk 4]
```diff
@@ -609,6 +609,34 @@ public function lockRevisions(ContentType $contentType, \DateTime $until, $by, $
...
+        $qbSelect
+            ->select('s.id')
+            ->andWhere($qbSelect->expr()->eq('s.contentType', ':content_type'))
+            ->andWhere($qbSelect->expr()->eq('s.lockBy', ':by'))
+            ->andWhere($qbSelect->expr()->isNull('s.endTime'))
+            ->andWhere($qbSelect->expr()->eq('s.deleted', $qbSelect->expr()->literal(false)))
+            ->andWhere($qbSelect->expr()->eq('s.draft', $qbSelect->expr()->literal(false)))
+        ;
+
+        /** @var QueryBuilder $qbUpdate */
```
Comment (Davidmattei) [inline Repository/RevisionRepository.php]: please remove, $this->createQueryBuilder() has a correct doc block

File: Service/DataService.php [hunk 5]
```diff
@@ -2010,4 +2010,83 @@ public function createAndMapIndex(Environment $environment): void
...
+        foreach ($contentTypes as $contentType) {
+            $draftedRevisions = \array_merge($draftedRevisions, $this->getContentTypeDraftedRevisions($contentType));
+        }
+
+        return $draftedRevisions;
+    }
+
+    public function getContentTypeDraftedRevisions(ContentType $contentType): array
+    {
+        return $this->revRepository->findBy([
```
Comment (Davidmattei) [inline Service/DataService.php]: Move this to a custom repository function, with a good name. Suggestion findDraftsByContentType

File: Service/DataService.php [hunk 6]
```diff
@@ -2010,4 +2010,83 @@ public function createAndMapIndex(Environment $environment): void
...
+        $rows = 0;
+        $contentTypes = $this->contentTypeService->getAll();
+        foreach ($contentTypes as $contentType) {
+            $rows += $this->lockContentTypeRevisions($contentType, $until, $by);
+        }
+
+        return $rows;
+    }
+
+    public function lockContentTypeRevisions(ContentType $contentType, \DateTime $until, string $by): int
```
Comment (Davidmattei) [inline Service/DataService.php]: I don't get the $rows variable the lockRevisions is called once so just return that result. And maybe add return type def on lockRevisions method

File: Service/DataService.php [hunk 7]
```diff
@@ -2010,4 +2010,83 @@ public function createAndMapIndex(Environment $environment): void
...
+        $rows = 0;
+        $contentTypes = $this->contentTypeService->getAll();
+        foreach ($contentTypes as $contentType) {
+            $rows += $this->unlockContentTypeRevisions($contentType, $by);
+        }
+
+        return $rows;
+    }
+
+    public function unlockContentTypeRevisions(ContentType $contentType, string $by): int
```
Comment (Davidmattei) [inline Service/DataService.php]: remove $rows

File: Service/DataService.php [hunk 8]
```diff
@@ -2010,4 +2010,83 @@ public function createAndMapIndex(Environment $environment): void
...
+                EmsFields::LOG_USERNAME_FIELD => $by,
+                EmsFields::LOG_EXCEPTION_FIELD => $e,
+                EmsFields::LOG_ERROR_MESSAGE_FIELD => $e->getMessage(),
+            ]);
+        }
+
+        return  $rows;
+    }
+
+    public function getAllContentTypesDraftedRevisions(): array
```
Comment (Davidmattei) [inline Service/DataService.php]: We just need a findAllDrafts method on the revision repository?

File: Service/DataService.php [hunk 9]
```diff
@@ -2010,4 +2010,83 @@ public function createAndMapIndex(Environment $environment): void
...
         ]);
     }
+
+    public function lockAllContentTypesRevisions(\DateTime $until, string $by): int
+    {
+        $rows = 0;
+        $contentTypes = $this->contentTypeService->getAll();
+        foreach ($contentTypes as $contentType) {
+            $rows += $this->lockContentTypeRevisions($contentType, $until, $by);
+        }
```
Comment (dameert) [inline Service/DataService.php]:
```suggestion
        foreach ($this->contentTypeService->getAll(); as $contentType) {
            $rows += $this->lockContentTypeRevisions($contentType, $until, $by);
        }
```

File: Command/UnlockRevisionsCommand.php [hunk 10]
```diff
@@ -0,0 +1,171 @@
...
+
+        $this->user = $input->getArgument(self::ARGUMENT_USER);
+    }
+
+    protected function execute(InputInterface $input, OutputInterface $output): int
+    {
+        try {
+            if ($this->all) {
+                $count = $this->dataService->unlockAllContentTypesRevisions($this->user);
+            } else {
```
[the remark this answers is not in the document]
Comment (coppee) [reply Command/UnlockRevisionsCommand.php]: @dameert interesting case (for me) Here, how do you handle the else case ? Duplicate "io->success" and return ?

File: Command/UnlockRevisionsCommand.php [hunk 11]
```diff
@@ -0,0 +1,171 @@
...
+    private $user;
+    /** @var ContentType */
+    private $contentType;
+    /** @var bool */
+    private $all;
+
+    private const ARGUMENT_USER = 'user';
+    private const ARGUMENT_CONTENT_TYPE = 'content-type';
+    private const OPTION_ALL = 'all';
+    private const OPTION_STRICT = 'strict';
```
[the remark this answers is not in the document]
Comment (coppee) [reply Command/UnlockRevisionsCommand.php]: enhancement #283 (I will make a draft in a new PR)

File: Repository/RevisionRepository.php [hunk 12]
```diff
@@ -628,4 +661,27 @@ public function findAllLockedRevisions(ContentType $contentType, string $lockBy,
 
         return new Paginator($qb->getQuery());
     }
+
+    public function findDraftsByContentType(ContentType $contentType): array
+    {
+        return $this->findBy([
```
Comment (Davidmattei) [inline Repository/RevisionRepository.php]: I don't like the magic doctrine findBy function, it creates a depend that we don't want.
Can you write the query using a queryBuilder

File: Repository/RevisionRepository.php [hunk 13]
```diff
@@ -628,4 +661,27 @@ public function findAllLockedRevisions(ContentType $contentType, string $lockBy,
...
+            'deleted' => false,
+            'endTime' => null,
+        ], [
+            'modified' => 'asc'
+        ]);
+    }
+
+    public function findAllDrafts(): array
+    {
+        return $this->findBy([
```
Comment (Davidmattei) [inline Repository/RevisionRepository.php]: see above comment

File: Repository/RevisionRepository.php [hunk 14]
```diff
@@ -609,6 +602,46 @@ public function lockRevisions(ContentType $contentType, \DateTime $until, $by, $
         return $qbUpdate->getQuery()->execute();
     }
 
+    public function lockAllRevisions(\DateTime $until, string $by): int
+    {
+        return $this->lockRevisions(null, $until, $by, true, false);
+    }
+
+    public function unlockRevisions(?ContentType $contentType, string $by): int
```
Comment (Davidmattei) [inline Repository/RevisionRepository.php]: nullable parameters should be at the end

File: Repository/RevisionRepository.php [hunk 15]
```diff
@@ -558,30 +558,23 @@ public function deleteRevisions(ContentType $contentType = null)
...
         $qbSelect
             ->select('s.id')
-            ->andWhere($qbSelect->expr()->eq('s.contentType', ':content_type'))
             ->andWhere($qbSelect->expr()->isNull('s.endTime'))
             ->andWhere($qbSelect->expr()->eq('s.deleted', $qbSelect->expr()->literal(false)))
             ->andWhere($qbSelect->expr()->eq('s.draft', $qbSelect->expr()->literal(false)))
         ;
 
+        if (null !== $contentType) {
+            $params['content_type'] = $contentType;
```
Comment (Davidmattei) [inline Repository/RevisionRepository.php]: you can the QueryBuilder setParameter() function multiple times, that way you don't need to keep the params array.

File: Repository/RevisionRepository.php [hunk 16]
```diff
@@ -558,30 +558,23 @@ public function deleteRevisions(ContentType $contentType = null)
...
-     * @param \DateTime $until
-     * @param string $by
-     * @param bool $force
-     *
-     * @param bool $id
-     * @return int affected rows
-     * @throws \Exception
-     */
-    public function lockRevisions(ContentType $contentType, \DateTime $until, $by, $force = false, $id = false)
+    public function lockRevisions(?ContentType $contentType, \DateTime $until, $by, $force = false, $id = false): int
```
Comment (Davidmattei) [inline Repository/RevisionRepository.php]: nullable parameters at the end, if this does creates any BC breaks ....

File: Command/TransformContentTypeCommand.php [hunk 17]
```diff
@@ -13,24 +15,27 @@
...
+    /** @var ContentType */
+    private $contentType;
+    /** @var string */
+    private $user;
 
-    const ARGUMENT_CONTENTTYPE_NAME = 'contentTypeName';
+    const ARGUMENT_CONTENT_TYPE = 'contentType';
+    const ARGUMENT_USER = 'user';
     const OPTION_STRICT = 'strict';
+    const DEFAULT_USER = 'TRANSFORM_CONTENT';
```
Comment (Davidmattei) [inline Command/TransformContentTypeCommand.php]: please add access type, private, public, protected?

File: Repository/RevisionRepository.php (hunk 16 above)
Comment (coppee) [reply Repository/RevisionRepository.php]: Unfortunately the function was already use somewhere else before my refactor (my bad!)

Comment (Davidmattei) [reply Repository/RevisionRepository.php]: No problem!

File: Repository/RevisionRepository.php [hunk 18]
```diff
@@ -576,57 +576,86 @@ public function deleteRevisions(ContentType $contentType = null)
...
+        }
+
         if (!$force) {
             $qbSelect->andWhere($qbSelect->expr()->orX(
                 $qbSelect->expr()->lt('s.lockUntil', ':now'),
                 $qbSelect->expr()->isNull('s.lockUntil')
             ));
-
-            $params['now'] = new \DateTime();
+            $qbUpdate->setParameters(['now' => new \DateTime()]);
```
Comment (Davidmattei) [inline Repository/RevisionRepository.php]:
```suggestion
            $qbUpdate->setParameters('now', new \DateTime());
```
Shoiuld be setParameter, setParmaters will not do a merge

File: Repository/RevisionRepository.php [hunk 19]
```diff
@@ -576,57 +576,86 @@ public function deleteRevisions(ContentType $contentType = null)
...
         }
 
         if ($id) {
             $qbSelect->andWhere(
                 $qbSelect->expr()->eq('s.ouuid', ':content_id')
             );
-            $params['content_id'] = $id;
+            $qbUpdate->setParameters(['content_id' => $id]);
```
Comment (Davidmattei) [inline Repository/RevisionRepository.php]:
```suggestion
            $qbUpdate->setParameter('content_id', $id);
```

File: Repository/RevisionRepository.php [hunk 20]
```diff
@@ -576,57 +576,86 @@ public function deleteRevisions(ContentType $contentType = null)
...
+        $qbUpdate = $this->createQueryBuilder('r');
+        $qbUpdate
+            ->update()
+            ->set('r.lockBy', ':by')
+            ->set('r.lockUntil', ':until')
+            ->setParameters(['by' => $by, 'until' => $until]);
+
+        if (null !== $contentType) {
+            $qbSelect->andWhere($qbSelect->expr()->eq('s.contentType', ':content_type'));
+            $qbUpdate->setParameters(['content_type' => $contentType]);
```
Comment (Davidmattei) [inline Repository/RevisionRepository.php]:
```suggestion
            $qbUpdate->setParameter('content_type', $contentType);
```

File: Repository/RevisionRepository.php (hunk 14 above)
Comment (coppee) [reply Repository/RevisionRepository.php]: Unfortunately the function was already use somewhere else before my refactor (my bad!)

File: Repository/RevisionRepository.php (hunk 18 above)
Comment (coppee) [reply Repository/RevisionRepository.php]: (thanks for that)

**Pull request merged**